File: ScorpioProto/src/Commons/ScorpioWriter.cs
using System;
using System.IO;
using System.Text;
namespace Scorpio.Commons
{
    public class ScorpioWriter
    {
        MemoryStream stream;
        BinaryWriter writer;
        public ScorpioWriter()
        {
            stream = new MemoryStream();
            writer = new BinaryWriter(stream);
        }
        public void WriteBool(bool value)
        {
            writer.Write(value ? (sbyte)1 : (sbyte)0);
        }
        public void WriteInt8(sbyte value)
        {
            writer.Write(value);
        }
        public void WriteInt16(short value)
        {
            writer.Write(value);
        }
        public void WriteInt32(int value)
        {
            writer.Write(value);
        }
        public void WriteInt64(long value)
        {
            writer.Write(value);
        }
        public void WriteFloat(float value)
        {
            writer.Write(value);
        }
        public void WriteDouble(double value)
        {
            writer.Write(value);
        }
        public void WriteString(string value)
        {
            ScorpioUtil.WriteString(writer, value);
        }
        public void WriteBytes(byte[] value)
        {
            writer.Write((int)value.Length);
            writer.Write(value);
        }
        public byte[] ToArray()
        {
            stream.Position = 0;
            return stream.ToArray();
        }
    }
}

File: ScorpioProto/Java/src/ScorpioProto/Commons/ScorpioWriter.java
package ScorpioProto.Commons;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
public class ScorpioWriter {
	private static final int MAX_LENGTH = 1024 * 1024;
	ByteBuffer writer;
	public ScorpioWriter() {
		writer = ByteBuffer.allocate(MAX_LENGTH).order(ByteOrder.LITTLE_ENDIAN);
	}
	public void WriteBool(boolean value) {
		writer.put(value ? (byte) 1 : (byte) 0);
	}
	public void WriteInt8(byte value) {
		writer.put(value);
	}
	public void WriteInt16(short value) {
		writer.putShort(value);
	}
	public void WriteInt32(int value) {
		writer.putInt(value);
	}
	public void WriteInt64(long value) {
		writer.putLong(value);
	}
	public void WriteFloat(float value) {
		writer.putFloat(value);
	}
	public void WriteDouble(double value) {
		writer.putDouble(value);
	}
	public void WriteString(String value) {
		try {
			if (value == null || value == "") {
				writer.put((byte) 0);
			} else {
				writer.put(value.getBytes("utf-8"));
				writer.put((byte) 0);
			}
		} catch (Exception e) {
		}
	}
	public void WriteBytes(byte[] value) {
		writer.putInt((int) value.length);
		writer.put(value);
	}
	public byte[] ToArray() {
		writer.flip();
		writer.rewind();
		byte[] bytes = new byte[writer.limit()];
		writer.get(bytes);
		return bytes;
	}
	public void Close() {
	}
}